engineapi/proxy.go

```diff
@@ -172,27 +404,164 @@ type Proxy struct {
 
 	grpcClient *imclient.ProxyClient
 
-	lock *sync.RWMutex
+	lock sync.Mutex
+
+	inUse     int
+	inUseLock sync.Mutex
+
+	nextResync *time.Time
+	resyncLock sync.Mutex
+
+	nextGC  *time.Time
+	gcRetry int
+	gcLock  sync.Mutex
+}
+
+type EngineClientProxy interface {
+	EngineClient
+	BackupTargetBinaryClient
+
+	GetLogger() logrus.FieldLogger
+
+	IsGRPC() bool
+	GetConnectionState() connectivity.State
+
+	Start(int, *longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore, chan struct{}) error
+	Stop() error
+
+	ShouldCreate(*longhorn.InstanceManager) bool
+	ShouldDelete() bool
+
+	SetNextResync(clock.Clock, time.Duration)
+	ShouldResync(clock.Clock) bool
+
+	SetNextGC(clock.Clock)
+	GetNextGC() *time.Time
+	GetGCRetry() int
+
+	SetInUse(bool)
+	GetInUse() int
```

Comment:
Suggest not adding these internal methods to the interface, so we don't need to make EngineBinary implemented these as well.

Reply:
Addressed.